site/content/how-to/traffic-management/snippets-filters.md

```diff
@@ -0,0 +1,345 @@
+and can be used in cases where Gateway API resources or NGINX extensions don't apply.
+
+## Disadvantages of Snippets
+
+Snippets are configured using the `SnippetsFilter` API, but are disabled by default due to their complexity and security implications.
+
+To use Snippets, set the `nginxGateway.snippetsFilters.enable` command line argument to true.
+
+Snippets have the following disadvantages:
+
+- *Complexity*. Snippets require you to:
+  - Understand NGINX Configuration primitives and implement a correct NGINX configuration.
+  - Understand how NGINX Gateway Fabric generates NGINX configuration so that a Snippet doesn’t interfere with the other features in the configuration.
+- *Decreased robustness*. An incorrect Snippet can invalidate NGINX configuration, causing reload failures. Until the snippet is fixed, it will prevent any new configuration updates, including updates for the other Gateway resources.
+- *Security implications*. Snippets give access to NGINX configuration primitives, which are not validated by NGINX Gateway Fabric. For example, a Snippet can configure NGINX to serve the TLS certificates and keys used for TLS termination for Gateway resources.
+
+{{< note >}} If the NGINX configuration includes an invalid Snippet, NGINX will continue to operate with the last valid configuration. {{< /note >}}
+
+Due to the described disadvantages of Snippets, we recommend exhausting all other means of NGINX configuration through Gateway API resources,
+[first-class policies]({{< relref "overview/custom-policies.md" >}}), and other NGINX extensions before using Snippets.
+
+{{< note >}} If you end up using Snippets and run into situations where an NGINX directive fails to be applied, please create an issue in the
+[NGINX Gateway Fabric Github repository](https://github.com/nginxinc/nginx-gateway-fabric). {{< /note >}}
+
+## Setup
+
+- [Install]({{< relref "/installation/" >}}) NGINX Gateway Fabric.
+- Save the public IP address and port of NGINX Gateway Fabric into shell variables:
+
+   ```text
+  GW_IP=XXX.YYY.ZZZ.III
+  GW_PORT=<port number>
+  ```
+
+  {{< note >}}In a production environment, you should have a DNS record for the external IP address that is exposed, and it should refer to the hostname that the gateway will forward for.{{< /note >}}
+
+- Create the coffee and tea example applications:
+
+  ```yaml
+  kubectl apply -f https://raw.githubusercontent.com/nginxinc/nginx-gateway-fabric/v1.4.0/examples/snippets-filter/app.yaml
+  ```
+
+- Create a Gateway:
+
+  ```yaml
+  kubectl apply -f https://raw.githubusercontent.com/nginxinc/nginx-gateway-fabric/v1.4.0/examples/snippets-filter/gateway.yaml
+   ```
+
+- Create HTTPRoutes for the coffee and tea applications:
+
+  ```yaml
+  kubectl apply -f https://raw.githubusercontent.com/nginxinc/nginx-gateway-fabric/v1.4.0/examples/snippets-filter/httproutes.yaml
+   ```
+
+- Test the configuration:
```

Comment:
I don't think the flow of the doc should be "deploy half the config, see that it doesn't work, now apply the other half".

We should focus on the happy path of deploying all of the necessary items to achieve a goal with snippets filters. At the end, we can talk about common errors (such as filter not being attached, what the result might be, and how to fix it).

Reply:
I didn't plan for it to be this way. I actually thought that an HTTPRoute that references a SnippetsFilter that doesn't exist would work normally, just without the SnippetsFilter settings. 

But it was only when I actually tested it out that I realized it error'd and I had to look into the SnippetsFilter proposal to figure out why. It's not specified anywhere, NGF doesn't error, and the SF itself has no errors, the only way to see the error is through describing the HTTPRoute. So I figured that if I had this issue, others would too, and it would be beneficial to highlight this important implementation feature. This guide intentionally breaks the norm to our guides setup phase.

I was also struggling to figure out a way to show the difference before and after snippetsfilters were applied. I could have them apply an HTTPRoute that doesn't reference the snippetsFilters, show traffic works, then apply a new HTTPRoute that references the SF and apply the SF, but that felt excessive. 

Open to change it though if you feel strongly that it should be changed.

@ADubhlaoich do you have any thoughts?

Reply:
I agree with @sjberman that starting the reader in a partially functional state feels very odd.

One thing that stuck out to me on a second pass is that the document goes from enabling Snippets immediately into an example for testing them.

I might be wrong, but from a quick look I can't actually find any other documentation on rate limiting in NGF, so we warn people that Snippets should be their last resort but don't provide any explanation how to do so with the suggest methods.

The caveat you mentioned about both NGF and the SnippetsFilter not erroring feels like a bug: it might warrant a warning somewhere, but, it feels odd to provide guidance around a design problem.

With this in mind, I'd probably try to give users a complete, functional example without the "see that it's broken step".

Here's my suggestion for an alternative page layout:

```
## Overview
## Disadvantages of Snippets
## Enable SnippersFilter
(This would essentially be the top of the "set-up" section now)
## Create an example application
(Include everything here needed to simply have a functional example. Nothing involving rate limiting yet: add two targets)
### Add rate limiting using <> (Whatever the simplest solution is)
(Add rate limiting to one target using a "regular" resource")
### Add rate limiting using SnippetsFilter
(Add rate limiting to one target using a SnippetsFilter)
### Test the rate limit
(Testing both targets the same way)
## Conclusions
## See also
```

Add describes at the end of each of the "Add rate limiting" sections so the reader can verify the changes they've made.

We give the non-Snippets solution first so the reader can see how to add rate limiting in a... Gateway "native" way, before presenting them with the Snippets solution we've already advised them away from.

Reply:
Correct me if I'm wrong @sjberman, but rate limiting is not a feature accessible through other means. So there are no other alternatives. 

SnippetsFilters give the users access to so much and so many NGINX features that currently aren't accessible which is such an upside. But we say it should be their last resort because it requires such extensible NGINX knowledge which the whole point of NGF is configure through other stuff so the user doesn't really need to touch NGINX conf.

Reply:
If that's the case, then "Add rate limiting using Snippets" could be its own how-to document: some alternative functionality could be chosen for this particular guide to narrow the focus onto something simpler.

I'm just throwing some ideas at the wall: I do stand by my prior suggestion of splitting the document into sections with one specific focus at a time.

Reply:
@ADubhlaoich made some changes. Removed the troubleshooting part, had the user deploy just a regular HTTPRoute without mention of SnippetsFilters, then later they apply the SnippetsFilter and modify the HTTPRoute. 

>If that's the case, then "Add rate limiting using Snippets" could be its own how-to document: some alternative functionality could be chosen for this particular guide to narrow the focus onto something simpler.

I think rate limiting is a simple feature which showcases Snippets in a good way. Since there are so many features SnippetsFilters opens up, we're not targeting and showcasing rate limiting as if we were to do that, we would also need to create docs for the other endless possibilities.

I based my document off of https://docs.nginx.com/nginx-gateway-fabric/how-to/traffic-management/client-settings where it just showcases a feature now enabled through the clientsettingspolicy, and doesn't really include any alternative method comparison.

I'm pushing back a little on your recommendations but I would like to hear your thoughts on the current changes.

Reply:
Current changes seem great - and keeping the scope of the PR to the original example is for the best.